### PR TITLE
Support collections of albums

### PR DESCRIPTION
**Pull request by Jaza**
This is a pretty big change, to make AWSPics optionally generate a site with a three-level hierarchy of:

- home page (that lists collections)
  - collection page (that lists albums)
    - album page (that lists pictures)

See this demo site with collections enabled, that I have set up at:

https://awspicstest.greenash.net.au/
username: `username`
password: `password`

Such a site is generated only if the new config option `groupAlbumsIntoCollections=true` is set in `config.json` . By default, AWSPics still generates a site with the usual two-level hierarchy of:

- home page (that lists albums)
  - album page (that lists pictures)

See this demo site with collections disabled (and with this change deployed, just to demonstrate that everything looks the same as it did before if using default config), that I have set up at:

https://awspicstest2.greenash.net.au/
username: `username`
password: `password`

Various other changes included in this PR:

- Allow the default `pics/original/` S3 object path to be overridden via `picsOriginalPath`
- Allow forcing the site builder to sort albums in asc / desc order via `albumSort`
- Allow forcing the site builder to sort pictures in asc / desc order via `pictureSort`
- Allow forcing the site builder to sort collections in asc / desc order via `collectionSort`
- Allow generating HTML markup indented with spaces instead of tabs via `spacesInsteadOfTabs`
- By default, show a "Design: HTML5 UP" credits link in bottom-right corner of home page, which can be turned off via `hideHomePageCredits`
- Allow the link in bottom-right corner of home page to be set to any custom text via `homePageCreditsOverride`
- Allow a `cover_image` to be specified in `metadata.yml` (previously only `title` could be specified there), for any album or collection, to force a particular image to be used when listing that album or collection: for an album, just specify any `image123.jpg` in that album; for a collection, specify any `album111/image222.jpg` in that collection (if no `cover_image` is specified, then it defaults to the first image in the album or collection, as usual)
- New script `generate_random_collections` which is analogous to the existing `generate_random_albums`
- Fixed both of the above scripts to have the `--sse` option for uploading to S3 with server-side encryption (otherwise the scripts fail with errors)
- Moved all HTML markup out of the site builder JS code and into template files
- Split up the site builder code into smaller files
- Updated the site builder unit tests to reflect all these changes

Feedback welcome!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,7 +3,7 @@
 > An AWS CloudFormation stack to run a serverless password-protected photo
   gallery
 
-**Demo:** <https://awspics.net>  
+**Demo:** <https://awspics.net>
```
Comment (jpsim) [inline README.md:6]: These trailing spaces were intentional to have Credentials on a new line.

File: resize/index.js [hunk 2]
```diff
@@ -4,6 +4,18 @@ var AWS = require("aws-sdk");
...
+
+const DEFAULT_PICS_ORIGINAL_PATH = 'pics/original/';
+
+
+function getPicsOriginalPath() {
+  if (process.env.PICS_ORIGINAL_PATH) {
+    return process.env.PICS_ORIGINAL_PATH;
+  }
+
+  return DEFAULT_PICS_ORIGINAL_PATH;
```
Comment (jpsim) [inline resize/index.js:16]: Does JS let you do something like this?

```suggestion
  if (process.env.PICS_ORIGINAL_PATH) {
    return process.env.PICS_ORIGINAL_PATH;
  }

  return process.env.PICS_ORIGINAL_PATH || DEFAULT_PICS_ORIGINAL_PATH;
```

**Review by jpsim — APPROVED** [review]
This is really cool, I love how you've cleaned up and organized a lot of the project.

File: README.md (hunk 1 above)
Comment (Jaza) [reply README.md:6]: Shouldn't rely on trailing spaces for this, or for anything else - trailing spaces are discouraged, and many people (including but not limited to me!) have their editors set to strip all trailing spaces when editing files.

I've added another newline between Demo and Credentials, and from a quick test, that seems to make GitHub render them on separate lines.

File: resize/index.js (hunk 2 above)
Comment (Jaza) [reply resize/index.js]: Yes, indeed it does - I think I had considered writing it like that, but didn't trust my JS knowledge enough, so I chickened out and wrote it more verbosely. Anyway, all tests still pass with it written as a one-liner, therefore it must be equivalent. I've committed your suggestion - thanks!

File: README.md (hunk 1 above)
Comment (jpsim) [reply README.md:6]: I didn't _make_ the Markdown spec, I just _write_ Markdown 😆 

Your latest change isn't equivalent to what was there before, but I don't feel strongly:

image

**Comment by Jaza** [comment]
Oops, I forgot to mention that this PR also allows specifying a `cover_image` in `metadata.yml` files, have updated the PR description to include that.

**Comment by Jaza** [comment]
Glad you like it @jpsim - this was the only big PR I was planning to submit for AWSPics, there are a few more minor tweaks I'd like to make, but no more massive new features or mega refactoring. :sunglasses: 

Well done again on building AWSPics. After all my unit test writing and feature building, I only actually tried deploying it for the first time a few days ago, and I was pleased to discover that it was a pretty smooth and painless process, got the first test site up easily enough, just by following the README instructions.

File: README.md (hunk 1 above)
Comment (Jaza) [reply README.md:6]: Oh yeah, true ... sorry about that.

Seems that GitHub renders Markdown slightly differently to other renderers, too (although I already vaguely knew that). For example, if I paste the text in to dillinger.io with neither the trailing spaces nor the extra newline, it renders Demo and Credentials on separate lines.

readme-in-dillinger

Comment (jpsim) [reply README.md:6]: No problem at all, just sharing why I went with the trailing spaces approach originally.

**Pull request merged**